conn.go

```diff
@@ -466,40 +466,122 @@ func (c *Conn) handleData(arg string) {
 	// We have recipients, go to accept data
 	c.WriteResponse(354, EnhancedCode{2, 0, 0}, "Go ahead. End your data with <CR><LF>.<CR><LF>")
 
-	var (
-		code         int
-		enhancedCode EnhancedCode
-		msg          string
-	)
+	defer c.reset()
+
+	if c.server.LMTP {
+		c.handleDataLMTP()
+		return
+	}
+
 	r := newDataReader(c)
-	err := c.Session().Data(r)
+	code, enhancedCode, msg := toSMTPStatus(c.Session().Data(r))
 	io.Copy(ioutil.Discard, r) // Make sure all the data has been consumed
-	if err != nil {
-		if smtperr, ok := err.(*SMTPError); ok {
-			code = smtperr.Code
-			enhancedCode = smtperr.EnhancedCode
-			msg = smtperr.Message
-		} else {
-			code = 554
-			enhancedCode = EnhancedCode{5, 0, 0}
-			msg = "Error: transaction failed, blame it on the weather: " + err.Error()
+	c.WriteResponse(code, enhancedCode, msg)
+
+}
+
+type statusCollector struct {
+	mapLock sync.Mutex
+
+	// Contains map from recipient to list of channels that are used for that
+	// recipient.
+	//
+	// First SetStatus call uses first channel, second - second, etc. Channels
+	// that are already used are set to nil (otherwise we can accidentally
+	// reuse channel that was consumed by handleDataLMTP already).
+	statusMap map[string][]chan error
```

Comment:
I'm not sure I understand why we need such a complicated data structure. Wouldn't a single channel be enough, with a `map[string]bool` on the receiver side to keep track of recipients we got status for?

Reply:
The same recipient can be specified N times and we should send N status codes for it, possibly different ones. Though I don't remember why I used `[]chan error`, perhaps it should be a single channel.

Reply:
Maybe then use a `map[string]int` to keep track of how many status replies we've sent for a recipient? (And still only one channel for everybody)

Reply:
What's about order then? Who should be responsible for making sure responses are sent in the correct order (same as RCPT TO order)?

Reply:
If we consider backends that do delivery for different recipients in parallel, it might be useful for go-smtp to keep track of the order instead (what this PR does, btw).

`map[string]chan error` alongside `[]chan error` looks good enough for me to solve this. With a single channel for everybody we need to do something when we receive statuses in a different order. Probably by putting statuses we can't send now into map[string][]error and using them later. What do you think?